text/orderer-v3.md

```diff
@@ -0,0 +1,320 @@
+- **Kafka**: Several ordering nodes relay transactions into an [Apache Kafka](https://kafka.apache.org/) system,
+  pull transactions back in the same order and then cut them into identical blocks via synchronization messages also sent through Kafka.
+- **Raft**: Several ordering nodes run the [Raft](https://raft.github.io/) protocol by embedding the [Raft library of etcd](https://github.com/etcd-io/etcd/tree/main/raft).
+
+Apart from the official Hyperledger Fabric consensus implementation, there is also a [Fabric fork](https://github.com/SmartBFT-Go/fabric) of a [BFT library](https://github.com/SmartBFT-Go/consensus) called SmartBFT.
+
+A BFT consensus protocol might have a different approach for onboarding and state snapshot replication compared to a CFT consensus protocol.
+In particular, a state snapshot sent between nodes in CFT is always assumed to be trusted, while in BFT it is not the case. 
+While it is tempting to have common code that performs state snapshot transfer that caters to both adversarial models, 
+such code is most likely to be complex and hard to follow and test.
+
+Each orderer type requires Fabric to understand how to initialize instances of it from the consensus metadata in the configuration blocks.
+However, this makes it impossible for external contributors to use their own consensus implementation without either forking Fabric entirely, 
+or making intrusive changes in the official Fabric orderer and peer code base.
+
+Undoubtedly, an alternative architecture which doesn't require Fabric codebase changes to add support for new 
+consensus protocols would be a sought-after property.
+
+# Identifying building blocks
+[building-blocks]: #identifying-building-blocks
+
+Apart from consensus, which agrees on orders of transaction, the orderer contains infrastructure that manages various aspects of its lifecycle and operations:
+
+ - **Chain management**: Spawns instances of the consensus protocols per channel and dispatches messages and operations to the right 
+   consensus instance.
+ - **Block replication service**: Server-side API that sends blocks upon request. 
+ - **Block synchronization utility**: Pulls blocks from other ordering service nodes and appends them to the local ledger. 
+ - **Communication service**: Authenticates client connections, receives messages from remote nodes and sends message to remote nodes.
+ - **Ledger subsystem**: Appends blocks to persistent storage and retrieves blocks by their index.
+ - **Channel participation service**: Manages local addition and removal of ledgers per channel.
+ - **Transaction filtering**: Enforces access control and validity checks on transactions both received from clients or from nodes.
+ - **Membership Service Provider subsystem**: Classifies identities into principals and drives policy authorization. 
+
+In this RFC, it is proposed that these building blocks be made consensus independent, and the building blocks that cannot be consensus independent will be 
```

Comment:
> In this RFC, it is proposed that these building blocks be made consensus independent,

I think, some of these building blocks map to the orderer specific common code which is present under https://github.com/hyperledger/fabric/tree/main/orderer/common and fabric core specific common code(deliver service) https://github.com/hyperledger/fabric/tree/main/common/deliver which is common for both peer & orderer node. 

The common implementation(like deliver service) for both core fabric & orderering service where we need to handle the inter dependencies.

Reply:
That is correct, a lot of the code is indeed independent, not not all of it. 

For example, onboarding which is in **orderer/common**, currently [imports Raft](https://github.com/hyperledger/fabric/blob/main/orderer/common/onboarding/onboarding.go#L110).

Reply:
I think it is worth highlighting which of the building blocks can largely be re-used from v2.x. And for the building blocks that require significant changes, what is the nature of those changes.